tabular/src/autogluon/tabular/predictor/predictor.py

```diff
@@ -2507,6 +2526,112 @@ def fit_weighted_ensemble(self, base_models: list = None, name_suffix='Best', ex
+                           f'and thus is invalid for decision threshold calibration. '
+                           f'Falling back to `decision_threshold=0.5`.')
+            return 0.5
+
+        if model == 'best':
+            # FIXME: Incorrect if 'val' and model is refit_full
+            model = self.get_model_best()
+
+        if data is None:
+            y_pred_proba_multi_val = self.predict_proba_multi(inverse_transform=False, as_multiclass=False)
+            y_pred_proba_val = y_pred_proba_multi_val[model]
+            val_data_source = 'val' if trainer.has_val else 'train'
+            _, y_val = self.load_data_internal(data=val_data_source, return_X=False, return_y=True)
+        else:
+            y_pred_proba_multi_val = self.predict_proba_multi(data=data, inverse_transform=False, as_multiclass=False)
+            y_pred_proba_val = y_pred_proba_multi_val[model]
+            y_val = self.transform_labels(labels=data[self.label])
+
+        if isinstance(decision_thresholds, int):
+            num_checks = decision_thresholds
+            decision_thresholds = [i / (num_checks - 1) for i in range(num_checks)]
+        best_score_val = None
+        best_threshold = None
+
+        y_pred_val = get_pred_from_proba(
+            y_pred_proba=y_pred_proba_val.values,
+            problem_type=self.problem_type,
+            decision_threshold=0.5,
+        )
+        # TODO: Avoid calling like this, re-use logic that works with weights + extra args
+        score_val_baseline = metric(y_true=y_val, y_pred=y_pred_val)
+
+        if verbose:
+            logger.log(20, f'\nOptimizing Metric: {metric.name}')
+        for decision_threshold in decision_thresholds:
+            extra_log = ''
+            y_pred_val = get_pred_from_proba(
+                y_pred_proba=y_pred_proba_val.values,
+                problem_type=self.problem_type,
+                decision_threshold=decision_threshold,
+            )
+            # TODO: Avoid calling like this, re-use logic that works with weights + extra args
+            score_val = metric(y_true=y_val, y_pred=y_pred_val)
+
+            if best_score_val is None or score_val > best_score_val:
+                best_threshold = decision_threshold
+                best_score_val = score_val
+                extra_log = '\t| NEW BEST'
+            elif best_score_val == score_val:
+                # If the new threshold is closer to 0.5 than the previous threshold, prioritize it.
+                if abs(decision_threshold - 0.5) < abs(best_threshold - 0.5):
+                    best_threshold = decision_threshold
+                    best_score_val = score_val
+                    extra_log = '\t| NEW BEST (Tie, using threshold that is closer to 0.5)'
+
+            if verbose:
+                logger.log(15, f'\tthreshold: {decision_threshold:.3f}\t| val: {score_val:.4f}{extra_log}')
```

Comment:
why don't we use bayesopt here. Something like this:
```python
from bayes_opt import BayesianOptimization

def score(self, threshold):
    # Implement your scoring function here
    # Make sure to pay attention if this greater is better here
    pass

optimizer = BayesianOptimization(
    f=score,
    pbounds={'threshold': (0, 1)},
    verbose=2, # verbose = 1 prints only when a maximum is observed, verbose = 0 is silent
    random_state=1,
)

optimizer.maximize(init_points=0, n_iter=decision_thresholds)

# Result
print("Best threshold: ", optimizer.max['params']['threshold'])
print("Best score: ", optimizer.max['target'])
```

Reply:
We can leave this optimization for a future release, as the logic is quite fast already.